runtime/gc_base/IndexableObjectAllocationModel.cpp

```diff
@@ -351,39 +466,100 @@ MM_IndexableObjectAllocationModel::doubleMapArraylets(MM_EnvironmentBase *env, J
 		return NULL;
 	}
 
-	GC_SlotObject *slotObject = NULL;
-	uintptr_t count = 0;
+	/* determine how many bytes to allocate outside of the spine (in arraylet leaves) */
+	Assert_MM_true(_allocateDescription.getBytesRequested() >= _allocateDescription.getContiguousBytes());
+	uintptr_t bytesRemaining = _allocateDescription.getBytesRequested() - _allocateDescription.getContiguousBytes();
+
+	/* allocate leaf for each arraylet and attach it to its leaf pointer in the spine */
+	uintptr_t arrayoidIndex = 0;
+	Trc_MM_getSparseAddressAndDecommitLeaves_Entry(env->getLanguageVMThread(), spine, (void *)bytesRemaining, arrayletLeafCount, (void *)arrayletLeafSize);
+	while (0 < bytesRemaining) {
+		/* allocate the next arraylet leaf - leaves are allocated solely for the purpose of decommitting the memory later on in this function */
+		void *leaf = env->_objectAllocationInterface->allocateArrayletLeaf(env, &_allocateDescription,
+				_allocateDescription.getMemorySpace(), true);
+
+		/* if leaf allocation failed set the result to NULL and return */
+		if (NULL == leaf) {
+			/* spine and preceding arraylets are now floating garbage */
+			Trc_MM_allocateAndConnectNonContiguousArraylet_leafFailure(env->getLanguageVMThread());
+			_allocateDescription.setSpine(NULL);
+			spine = NULL;
+			break;
+		}
+
+		/* Disable region for reads and writes, since that'll be done through the contiguous double mapped region */
+		void *highAddress = (void *)((uintptr_t)leaf + arrayletLeafSize);
+		bool ret = extensions->heap->decommitMemory(leaf, arrayletLeafSize, leaf, highAddress);
+		if (!ret) {
+			Trc_MM_VirtualMemory_decommitMemory_failure(leaf, arrayletLeafSize);
+		}
+
+		arrayletLeaveAddrs[arrayoidIndex] = leaf;
+
+		/* refresh the spine -- it might move if we GC while allocating the leaf */
+		spine = _allocateDescription.getSpine();
+
+		bytesRemaining -= OMR_MIN(bytesRemaining, arrayletLeafSize);
+		arrayoidIndex += 1;
+	}
+
+
+	if ((NULL != spine)
+	) {
+		Assert_MM_true(_layout == GC_ArrayletObjectModel::InlineContiguous);
+		Assert_MM_true(indexableObjectModel->isVirtualLargeObjectHeapEnabled());
+		/* Number of arraylet leaves in the iterator must match the number of leaves calculated */
+		Assert_MM_true(arrayletLeafCount == arrayoidIndex);
 
-	while (NULL != (slotObject = arrayletLeafIterator.nextLeafPointer())) {
-		void *currentLeaf = slotObject->readReferenceFromSlot();
-		arrayletLeaveAddrs[count] = currentLeaf;
-		count++;
+		byteAmount = _dataSize;
+		void *virtualLargeObjectHeapAddress = extensions->largeObjectVirtualMemory->allocateSparseFreeEntryAndMapToHeapObject(spine, byteAmount);
+		if (NULL != virtualLargeObjectHeapAddress) {
+			indexableObjectModel->setDataAddrForContiguous((J9IndexableObject *)spine, virtualLargeObjectHeapAddress);
+		}
+		/* Free arraylet leaf addresses if dynamically allocated */
+		if (arrayletLeafCount > ARRAYLET_ALLOC_THRESHOLD) {
+			env->getForge()->free((void *)arrayletLeaveAddrs);
+		}
```

Comment:
we might need a special state for this regions (different from just plain ARRAYLET), such as ARRAYLET_DECOMMITED or just DECOMMITED